File: src/utils/source.cpp
#include "utils/source.hpp"
#include "utils/chars.hpp"

using namespace utils;

SourceText::SourceText(const std::string &source)
    : source(source) {
    this->calculateLines();
}

char SourceText::getCharAt(uint64_t index) const {
    if (index >= this->source.length())
        return '\0';
    
    return this->source[index];
}

std::optional<uint64_t> SourceText::getLineLength(uint64_t line_index) const {
    if (line_index >= this->lines.size())
        return std::nullopt;
    
    return this->lines[line_index].getUtf8Length();
}

std::optional<std::string> SourceText::getLineText(uint64_t line_index) const {
    if (line_index >= this->lines.size())
        return std::nullopt;
    
    Position line = this->lines[line_index];
    return this->source.substr(line.getAsciiStart(), line.getAsciiLength());
}

std::optional<std::string> SourceText::getTextFromPosition(const Position &position) const {
    if (position.getAsciiEnd() > this->source.length())
        return std::nullopt;
    
    return this->source.substr(position.getAsciiStart(), position.getAsciiLength());
}

void SourceText::getInfosForPosition(const Position &position, SourceInformations &infos) const {
    uint64_t index = 0;
    infos.success = false;

    for (const Position &line : this->lines) {
        if (position.getUtf8Start() >= line.getUtf8Start() && position.getUtf8Start() <= line.getUtf8End()) {
            infos.line_start = index;
            infos.left_offset = position.getUtf8Start() - line.getUtf8Start();
        }

        if (position.getUtf8End() >= line.getUtf8Start() && position.getUtf8End() <= line.getUtf8End()) {
            infos.line_end = index;
            infos.right_offset = line.getUtf8End() - position.getUtf8End();
            infos.success = true;
            return;
        }

        index++;
    }
}

void SourceText::calculateLines() {
    if (this->source.length() == 0)
        return;
    
    uint64_t old_ascii_pos = 0, old_utf8_pos = 0;
    uint64_t actual_ascii_pos = 0, actual_utf8_pos = 0;
    uint8_t utf8_length = getUtf8Length(this->source[0]);

    for (char c : this->source) {
        if (c == '\n') {
            lines.push_back(Position(old_ascii_pos, actual_ascii_pos - old_ascii_pos, old_utf8_pos, actual_utf8_pos - old_utf8_pos));
            old_ascii_pos = actual_ascii_pos + 1;
            old_utf8_pos = actual_utf8_pos + 1;
        }

        actual_ascii_pos++;
        utf8_length--;
        if (utf8_length == 0) {
            utf8_length = getUtf8Length(c);
            actual_utf8_pos++;
        }
    }

    lines.push_back(Position(old_ascii_pos, actual_ascii_pos - old_ascii_pos, old_utf8_pos, actual_utf8_pos - old_utf8_pos));
}

File: src/utils/source.rs
use crate::utils::position::Position;

pub struct SourceText {
    source: String,
    lines: Vec<Position>,
}

impl SourceText {
    pub fn init(source: String) -> SourceText {
        let lines = Self::calculate_lines(&source);
        SourceText { source, lines }
    }

    pub fn get_char_at(&self, index: usize) -> char {
        self.source.chars().nth(index).unwrap_or('\0')
    }

    pub fn get_line_length(&self, index: usize) -> Option<usize> {
        self.lines.get(index).map(|line| line.get_length())
    }

    pub fn get_line_text(&self, index: usize) -> Option<String> {
        self.lines.get(index).map(|line| {
            self.source.chars()
                .skip(line.get_start())
                .take(line.get_length())
                .collect()
        })
    }

    pub fn get_text_from_position(&self, position: &Position) -> String {
        self.source.chars()
            .skip(position.get_start())
            .take(position.get_length())
            .collect()
    }

    pub fn get_infos_for_position(&self, pos: &Position, line_start: &mut usize, line_end: &mut usize, left_offset: &mut usize, right_offset: &mut usize) {
        let mut i: usize = 0;
        for line in &self.lines {
            if pos.get_start() >= line.get_start() && pos.get_start() <= line.get_end() {
                *line_start = i;
                *left_offset = pos.get_start() - line.get_start();
            }

            if pos.get_end() >= line.get_start() && pos.get_end() <= line.get_end() {
                *line_end = i;
                *right_offset = line.get_end() - pos.get_end();
            }

            i += 1;
        }
    }

    fn calculate_lines(source: &String) -> Vec<Position> {
        let mut lines = vec![];
        if source.is_empty() {
            return lines;
        }

        let mut old_pos: usize = 0;
        let mut actual_pos: usize = 0;
        for c in source.chars() {
            if c == '\n' {
                lines.push(Position::init(old_pos, actual_pos - old_pos));
                old_pos = actual_pos + 1;
            }

            actual_pos += 1;
        }

        lines.push(Position::init(old_pos, actual_pos - old_pos));
        lines
    }
}


#[cfg(test)]
mod tests {
    use crate::utils::source::SourceText;

    #[test]
    fn empty_source() {
        let source = SourceText::init(String::new());

        assert_eq!(source.get_line_length(0), None);
        assert_eq!(source.get_char_at(0), '\0');
        assert_eq!(source.get_line_text(0), None);
    }
}